grpc/src/signer.rs

```diff
@@ -0,0 +1,399 @@
+//! Types related to signing transactions
+
+use std::fmt;
+
+use ::tendermint::chain::Id;
+use async_trait::async_trait;
+use celestia_proto::cosmos::tx::v1beta1::SignDoc;
+use k256::ecdsa::signature::Signer;
+use k256::ecdsa::VerifyingKey;
+use prost::{Message, Name};
+use signature::Keypair;
+use tendermint_proto::google::protobuf::Any;
+use tendermint_proto::Protobuf;
+
+use celestia_proto::cosmos::crypto::secp256k1;
+use celestia_types::state::auth::BaseAccount;
+use celestia_types::state::{
+    AccAddress, AuthInfo, Fee, ModeInfo, RawTx, RawTxBody, SignerInfo, Sum,
+};
+
+use crate::Result;
+
+/// ECDSA/secp256k1 signature used for signing transactions
+pub type DocSignature = k256::ecdsa::Signature;
+/// Signature error
+pub type SignatureError = k256::ecdsa::signature::Error;
+
+/// Signer capable of producing ecdsa signature using secp256k1 curve.
+#[async_trait]
+pub trait DocSigner: Send + Sync {
+    /// Try to sign the provided sign doc.
+    async fn try_sign(&self, doc: SignDoc) -> Result<DocSignature, SignatureError>;
+}
+
+#[async_trait]
+impl<T> DocSigner for T
+where
+    T: Signer<DocSignature> + Send + Sync + 'static,
+{
+    async fn try_sign(&self, doc: SignDoc) -> Result<DocSignature, SignatureError> {
+        let bytes = doc.encode_to_vec();
+        self.try_sign(&bytes)
+    }
+}
+
+/// Workaround for dispatching `DocSigner`
+pub struct DispatchedDocSigner(Box<dyn DocSigner>);
+
+impl DispatchedDocSigner {
+    /// Create a new signer
+    pub fn new<S>(signer: S) -> DispatchedDocSigner
+    where
+        S: DocSigner + 'static,
+    {
+        DispatchedDocSigner(Box::new(signer))
+    }
+}
+
+#[async_trait]
+impl DocSigner for DispatchedDocSigner {
+    async fn try_sign(&self, doc: SignDoc) -> Result<DocSignature, SignatureError> {
+        self.0.try_sign(doc).await
+    }
+}
+
+pub(crate) trait KeypairExt {
+    fn address(&self) -> AccAddress;
+}
+
+impl<T> KeypairExt for T
+where
+    T: Keypair,
+    T::VerifyingKey: Into<AccAddress>,
+{
+    fn address(&self) -> AccAddress {
+        self.verifying_key().into()
+    }
+}
+
```

Comment:
this seem to only be used [here](https://github.com/eigerco/lumina/pull/712/files#diff-8b2379cf48ef774d4757c1ecd32705abea258152d0c229b2b8afde6628f28681R427), no?

In that case I'd prefer removing this and just using `self.signer.pubkey.into()` there

Reply:
got carried away here, removed ;)